### PR TITLE
cloud: update dependencies on Azure.

### LINKED ISSUE #14609: Update the dependencies of Azure SDK.
## Development Task
As the title shows, current Azure SDK is too old (version is `0.1.0`) to develop new features on it, which should be updated to the latest version `0.11`.

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close #14609 

What's Changed:

```commit-message
Update the Azure SDK to latest version to support later developments.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None.
```

### CODE REVIEW

**Comment by LykxSassinator** [comment]
@Leavrth @3pointer PTAL, thx

File: components/cloud/azure/Cargo.toml [hunk 1]
```diff
@@ -6,20 +6,25 @@ publish = false
 
 [dependencies]
 async-trait = "0.1"
-azure_core = { version = "0.1.0", git = "https://github.com/Azure/azure-sdk-for-rust" }
-azure_identity = { version = "0.1.0", git = "https://github.com/Azure/azure-sdk-for-rust" }
-azure_storage = { version = "0.1.0", git = "https://github.com/Azure/azure-sdk-for-rust", default-features = false, features = ["account", "blob"] }
```
Comment (tonyxuqqi) [inline components/cloud/azure/Cargo.toml:11]: why remove "default-features = false features = ["account", "blob"]"

File: components/cloud/azure/src/azblob.rs [hunk 2]
```diff
@@ -443,9 +441,9 @@ impl ContainerBuilder for TokenCredContainerBuilder {
             {
                 let token_response = self.token_cache.read().unwrap();
                 if let Some(ref t) = *token_response {
-                    let interval = t.0.expires_on - Utc::now();
+                    let interval = (t.0.expires_on - OffsetDateTime::now_utc()).whole_minutes();
```
Comment (tonyxuqqi) [inline components/cloud/azure/src/azblob.rs:444]: Does it compile with OffsetDateTime? 
Also make sure it works when expires_on < OffsetDateTime::now_utc() by adding a test case.

File: components/cloud/azure/Cargo.toml (hunk 1 above)
Comment (LykxSassinator) [reply components/cloud/azure/Cargo.toml:11]: About the first question:
> why remove "default-features = false features = ["account", "blob"]".

In the latest version, `account` feature is already abandoned, related traits have been included in `azure_storage` and `azure_storage_blob`. The `default-features = false` should not be ignored and added now.

And for:
> Also why use the same dependent component with different names?

They are different libs released by `azure-sdk-for-rust` and maintained in the same repo.

**Comment by LykxSassinator** [comment]
@Leavrth The ut, uploading blocks to Blob in Azure emulator, has passed.
```Rust
Running unittests src/lib.rs (target/debug/deps/azure-c5d083c56c4a1f29)

running 1 test
test azblob::tests::test_azblob_storage ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.14s
```

**Comment by Leavrth** [comment]
backup and restore test in Azure emulator has passed.
```
$ ./br backup full -s "azure://test/full-cluster5?endpoint=http://127.0.0.1:10000/devstoreaccount1&account-name=devstoreaccount1&account-key=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==&access-tier=Hot" -u 127.0.0.1:2379
Detail BR log in /tmp/br.log.2023-04-20T17.51.05+0800 
Full Backup <----------------------------------------------------------------------------------------------> 100.00%
Checksum <-------------------------------------------------------------------------------------------------> 100.00%
[2023/04/20 17:53:47.877 +08:00] [INFO] [collector.go:73] ["Full Backup success summary"] [total-ranges=6883] [ranges-succeed=6883] [ranges-failed=0] [backup-checksum=23.484616848s] [backup-fast-checksum=20.084123ms] [backup-total-ranges=297] [DBCount=2] [TableCount=148] [total-take=2m42.231220714s] [total-kv=5382354] [total-kv-size=274.5GB] [average-speed=1.692GB/s] [backup-data-size(after-compressed)=16.71GB] [Size=16706656413] [BackupTS=440922083650174977]


$ ./br restore full -s "azure://test/full-cluster5?endpoint=http://127.0.0.1:10000/devstoreaccount1&account-name=devstoreaccount1&account-key=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==&access-tier=Hot" -u 127.0.0.1:2379
Detail BR log in /tmp/br.log.2023-04-20T17.59.54+0800 
Full Restore <---------------------------------------------------------------------------------------------> 100.00%
[2023/04/20 18:01:13.820 +08:00] [INFO] [collector.go:73] ["Full Restore success summary"] [total-ranges=6957] [ranges-succeed=6957] [ranges-failed=0] [split-region=14.244757911s] [restore-ranges=3463] [DBCount=2] [TableCount=108] [total-take=1m19.197843975s] [restore-data-size(after-compressed)=16.65GB] [Size=16651714058] [BackupTS=440922083650174977] [RestoreTS=440922222073217025] [total-kv=5302882] [total-kv-size=274.5GB] [average-speed=3.465GB/s]

$ ./br restore point --full-backup-storage "azure://test/full-cluster5?endpoint=http://127.0.0.1:10000/devstoreaccount1&account-name=devstoreaccount1&account-key=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==&access-tier=Hot" -s "azure://test/log-cluster5?endpoint=http://127.0.0.1:10000/devstoreaccount1&account-name=devstoreaccount1&account-key=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==&access-tier=Hot" -u 127.0.0.1:43762
Detail BR log in /tmp/br.log.2023-04-20T18.36.47+0800 
Full Restore <---------------------------------------------------------------------------------------------> 100.00%
[2023/04/20 18:36:53.630 +08:00] [INFO] [collector.go:73] ["Full Restore success summary"] [total-ranges=42] [ranges-succeed=42] [ranges-failed=0] [split-region=1.21486ms] [restore-ranges=20] [DBCount=2] [TableCount=17] [total-take=6.237140538s] [RestoreTS=440922802395545614] [total-kv=5052400] [total-kv-size=824.7MB] [average-speed=132.2MB/s] [restore-data-size(after-compressed)=359.2MB] [Size=359163435] [BackupTS=440922420673773569]
Restore Meta Files <---------------------------------------------------------------------------------------> 100.00%
Restore KV Files <-----------------------------------------------------------------------------------------> 100.00%
[2023/04/20 18:37:31.701 +08:00] [INFO] [collector.go:73] ["restore log success summary"] [total-take=38.070552945s] [restore-from=440922420673773569] [restore-to=440922788291411969] [restore-from="2023-04-20 18:12:32.497 +0800"] [restore-to="2023-04-20 18:35:54.847 +0800"] [total-kv-count=19150353] [total-size=3.511GB] [average-speed=92.23MB/s]

$ ./sync_diff_inspector --config ./config2.toml
A total of 9 tables need to be compared

Comparing the table structure of ``test`.`stock`` ... equivalent
Comparing the table structure of ``test`.`new_order`` ... equivalent
Comparing the table structure of ``test`.`orders`` ... equivalent
Comparing the table structure of ``test`.`order_line`` ... equivalent
Comparing the table structure of ``test`.`item`` ... equivalent
Comparing the table structure of ``test`.`warehouse`` ... equivalent
Comparing the table structure of ``test`.`district`` ... equivalent
Comparing the table structure of ``test`.`customer`` ... equivalent
Comparing the table structure of ``test`.`history`` ... equivalent
Comparing the table data of ``test`.`new_order`` ... equivalent
Comparing the table data of ``test`.`stock`` ... equivalent
Comparing the table data of ``test`.`orders`` ... equivalent
Comparing the table data of ``test`.`warehouse`` ... equivalent
Comparing the table data of ``test`.`district`` ... equivalent
Comparing the table data of ``test`.`item`` ... equivalent
Comparing the table data of ``test`.`order_line`` ... equivalent
Comparing the table data of ``test`.`customer`` ... equivalent
Comparing the table data of ``test`.`history`` ... equivalent
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
A total of 9 table have been compared and all are equal.
You can view the comparision details through '/tmp/sync-diff.output.2023-04-20T18.38.23+0800/sync_diff.log'
```

**Review by Leavrth — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]

File: components/cloud/azure/src/azblob.rs [hunk 3]
```diff
@@ -498,28 +494,44 @@ impl AzureStorage {
...
+                                .blob_service_client()
+                                .container_client(bucket)
+                                .clone(),
+                        ),
+                    }),
+                });
+            }
+            let account_name = config.get_account_name()?;
+            let storage_credentials = ConnectionString::new(&connection_string)
+                .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, format!("{}", e)))?
```
Comment (3pointer) [inline components/cloud/azure/src/azblob.rs:518]: maybe give a clear error message here

Comment (Leavrth) [reply components/cloud/azure/src/azblob.rs:518]: After this PR is merged, I would commit anothor PR to make azblob support `encryption-scope`, and by the way I will make enhancement for azblob.

Comment (LykxSassinator) [reply components/cloud/azure/src/azblob.rs:518]: Got, it will be updated in other prs.

File: components/cloud/azure/src/azblob.rs [hunk 4]
```diff
@@ -498,28 +494,44 @@ impl AzureStorage {
...
-            let bucket = (*config.bucket.bucket).to_owned();
-            let http_client = new_http_client();
-            let container_client = StorageAccountClient::new_connection_string(
-                http_client.clone(),
-                connection_string.as_str(),
-            )
-            .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, format!("{}", &e)))?
-            .as_storage_client()
-            .as_container_client(bucket);
+            #[cfg(test)]
```
Comment (tabokie) [inline components/cloud/azure/src/azblob.rs:500]: cfg(all())

File: components/cloud/azure/src/azblob.rs [hunk 5]
```diff
@@ -498,28 +494,44 @@ impl AzureStorage {
...
+            #[cfg(feature = "azurite")]
+            {
```
Comment (tabokie) [inline components/cloud/azure/src/azblob.rs:502]: What is this block for? And FYI `cfg(test)` doesn't work for dependencies (if you test this function in another crate the behavior will be different), it's weird to have behavior changing silently like this.

Comment (LykxSassinator) [reply components/cloud/azure/src/azblob.rs:502]: For testing the Azure blob with a third party tool Azurite(a tool to mock a local server for testing Azure SDK).

Comment (tabokie) [reply components/cloud/azure/src/azblob.rs:501]: I looked again, this feature wasn't even registered in Cargo.toml. I think they are just dead code.

That aside, for this change itself, you should probably just separate it into a new function to keep it clean. One of the confusions I have is, say you are using an "Azurite" specific builder here, what if the `parse_plaintext_account_url` is none before? It falls back to normal implementation, which doesn't make sense to me as an outsider. If it really needs to fallback, then you should add plenty of comments to explain why we can use normal implementation to connect to Azurite, but cannot when `parse_plaintext_account_url` is some.

**Pull request merged**